src/libs/Violations/ViolationsUtils.ts

```diff
@@ -78,6 +79,53 @@ const ViolationsUtils = {
                 if (!hasMissingTagViolation && !updatedTransaction.tag && policyRequiresTags) {
                     newTransactionViolations.push({name: CONST.VIOLATIONS.MISSING_TAG, type: 'violation'});
                 }
+            } else {
+                newTransactionViolations = newTransactionViolations.filter(
+                    (violation) => violation.name !== CONST.VIOLATIONS.SOME_TAG_LEVELS_REQUIRED && violation.name !== CONST.VIOLATIONS.TAG_OUT_OF_POLICY,
+                );
+
+                // We first get the errorIndexes for someTagLevelsRequired. If it's not empty, we puth SOME_TAG_LEVELS_REQUIRED in Onyx.
+                // Otherwise, we put TAG_OUT_OF_POLICY in Onyx (when applicable)
+                const errorIndexes = [];
+                for (let i = 0; i < policyTagKeys.length; i++) {
+                    const isTagRequired = policyTagList[policyTagKeys[i]].required ?? true;
+                    const isTagSelected = Boolean(selectedTags[i]);
+                    if (isTagRequired && (!isTagSelected || (selectedTags.length === 1 && selectedTags[0] === ''))) {
+                        errorIndexes.push(i);
+                    }
+                }
+                if (errorIndexes.length !== 0) {
+                    newTransactionViolations.push({
+                        name: CONST.VIOLATIONS.SOME_TAG_LEVELS_REQUIRED,
+                        type: 'violation',
+                        data: {
+                            errorIndexes,
+                        },
+                    });
+                } else {
+                    let hasInvalidTag = false;
+                    for (let i = 0; i < policyTagKeys.length; i++) {
+                        const selectedTag = selectedTags[i];
+                        const tags = policyTagList[policyTagKeys[i]].tags;
+                        const isTagInPolicy = Object.values(tags).some((tag) => tag.name === selectedTag && Boolean(tag.enabled));
+                        if (!isTagInPolicy) {
+                            newTransactionViolations.push({
+                                name: CONST.VIOLATIONS.TAG_OUT_OF_POLICY,
+                                type: 'violation',
+                                data: {
+                                    tagName: policyTagKeys[i],
+                                },
+                            });
+                            hasInvalidTag = true;
+                            break;
+                        }
+                    }
+                    if (!hasInvalidTag) {
+                        newTransactionViolations = reject(newTransactionViolations, {
+                            name: CONST.VIOLATIONS.TAG_OUT_OF_POLICY,
+                        });
```

Comment:
`filter` can be used here as well.

Reply:
There's no difference here since we're only calling reject once, so it should go over the array only once, right?

Reply:
Yes, but aren't we preferring native JS methods for filtering after TS migration?